### PR TITLE
DATAUP-259 Check child jobs

### PR DESCRIPTION
**Pull request by bio-boris**
# Description of PR purpose/changes

* Get the status of a job and the status of its children
* Tests are running

# Testing Instructions
- [x] Tests pass in travis and locally

### CODE REVIEW

File: execution_engine2.spec [hunk 1]
```diff
@@ -391,8 +391,22 @@
         funcdef check_job(CheckJobParams params) returns (JobState job_state) authentication required;
 
         /*
-            job_states - states of jobs
+            parent_job - state of parent job
+            job_states - states of child jobs
+            IDEA: ADD aggregate_states - count of all available child job states, even if they are zero
```
Comment (ialarmedalien) [inline execution_engine2.spec:396]: should this line be in here?

File: execution_engine2.spec [hunk 2]
```diff
@@ -391,8 +391,22 @@
...
+        */
+        typedef structure {
+            JobState parent_jobstate;
+            list<JobState> child_jobstates;
+        } CheckJobBatchResults;
+
+        /*
+            get current status of a parent job, and it's children, if it has any.
```
Comment (ialarmedalien) [inline execution_engine2.spec:404]: Punctuation police alert:
`it's` = it is
`its` = belonging to it

Comment (bio-boris) [reply execution_engine2.spec:404]: ok fixing it in next pr

File: execution_engine2.spec (hunk 1 above)
Comment (bio-boris) [reply execution_engine2.spec:396]: Fixing this in next PR

File: lib/execution_engine2/sdk/SDKMethodRunner.py [hunk 3]
```diff
@@ -319,6 +317,41 @@ def get_job_status_field(self, job_id, as_admin=False):
         """ Authorization Required: Read """
         return self.get_jobs_status().get_job_status(job_id=job_id, as_admin=as_admin)
 
+    def check_job_batch(
+        self,
+        parent_job_id,
+        check_permission=True,
```
Comment (ialarmedalien) [inline lib/execution_engine2/sdk/SDKMethodRunner.py:323]: Does `check_job_batch` get called elsewhere, other than in Impl file? The implementation file doesn't have any way of adding the `check_permission` param.

File: lib/execution_engine2/sdk/SDKMethodRunner.py [hunk 4]
```diff
@@ -319,6 +317,41 @@ def get_job_status_field(self, job_id, as_admin=False):
...
+        if child_job_ids:
+            child_job_states = self.get_jobs_status().check_jobs(
+                job_ids=child_job_ids,
+                check_permission=True,
+                exclude_fields=exclude_fields,
+                return_list=1,
+            )["job_states"]
+        return {
+            "parent_jobstate": parent_job_status,
+            "child_jobstates": child_job_states,
```
Comment (ialarmedalien) [inline lib/execution_engine2/sdk/SDKMethodRunner.py:352]: is this going to be an array of all job states, or will it be a mapping with job ID => job state?

Comment (bio-boris) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:352]: array of all job states

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 3 above)
Comment (bio-boris) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:323]: No. This is simply a copy of the way things are done for "check_job". We want the path from the IMPL to always check permissions. Do you want me to move it into the function definition instead?

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 4 above)
Comment (ialarmedalien) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:352]: So how are you going to know which job is in which state?

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 3 above)
Comment (ialarmedalien) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:323]: Easiest would probably be to add the param to the function call in the Impl file.

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 4 above)
Comment (bio-boris) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:352]: @ialarmedalien It is a list of job states, not a list of job statuses. You can check the job[status] field to see the job status.

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 3 above)
Comment (bio-boris) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:323]: @ialarmedalien We don't want the user to be able to bypass permissions, but we've used it in the past before we had the permissions request mocking and before we had the permissions cache.

File: lib/execution_engine2/sdk/SDKMethodRunner.py (hunk 4 above)
Comment (ialarmedalien) [reply lib/execution_engine2/sdk/SDKMethodRunner.py:352]: Ah sorry, gotcha!

**Review by ialarmedalien — APPROVED** [review]
👍

**Pull request merged**